### PR TITLE
kube: make discovery selectors automagically handle updates

### PR DESCRIPTION
**Pull request by howardjohn**
For https://github.com/istio/istio/issues/44675

Previously, the discovery selectors were not integrated into our clients. This made it tricky to use -- each controller that used filters needed to subscribe to events on the filter and manually handle namespace add and deletes.

This is because the way we plug the filter into an informer doesn't handle dynamic filters (https://github.com/kubernetes/kubernetes/issues/117706).

This PR changes that, and makes our filtering dynamic. If a namespace is selected or de-selected, all items in that namespace are automatically triggered as Add/Deletes. Wit this, any controller can simply use the filter without needing special logic.

Additionally, to facilitate usage and simplify the lifecycle of the filter, the filter is now plugged into a `kube.Client`. This allows it to easily be used anywhere a `kube.Client` is (much like CRDWatch is today).

### CODE REVIEW

File: pilot/pkg/config/kube/crdclient/client.go [hunk 1]
```diff
@@ -395,6 +394,45 @@ func (cl *Client) addCRD(name string) {
...
+}
+
+func (f composedFilter) Filter(obj any) bool {
+	for _, filter := range f.extra {
+		if !filter(obj) {
+			return false
+		}
+	}
+	if f.filter != nil {
+		return f.filter.Filter(obj)
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:412]: It seems the orders matter, and extras take first. I think we need to add more comments about this

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -265,14 +254,7 @@ func NewController(kubeClient kubelib.Client, options Options) *Controller {
 		)
 	}
 
-	// always init for each cluster, otherwise different ns labels in different cluster may not take effect,
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:268]: it's part of the kube.Client which is constructed for each cluster

File: pkg/kube/kclient/client.go [hunk 3]
```diff
@@ -298,10 +306,52 @@ func newDelayedInformer[T controllers.ComparableObject](
...
+func applyDynamicFilter[T controllers.ComparableObject](filter Filter, ic *informerClient[T]) {
+	if filter.ObjectFilter != nil {
+		ic.filter = filter.ObjectFilter.Filter
+		filter.ObjectFilter.AddHandler(func(added, removed sets.String) {
+			ic.handlerMu.RLock()
+			defer ic.handlerMu.RUnlock()
+			if types.GetGVR[T]() == istiogvr.Namespace {
+				// Namespace is special; we query all namespaces
+				// Note: other cluster-scoped resources should just not use the filter
+				for _, item := range ic.ListUnfiltered(metav1.NamespaceAll, klabels.Everything()) {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/kclient/client.go:326]: We can either do a list+ filter or many Get calls. Probably many Get is a bit more efficient, I can change it

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 4]
```diff
@@ -890,14 +891,10 @@ func (c *Controller) AdditionalPodSubscriptions(
...
-			namespaces = c.opts.DiscoveryNamespacesFilter.GetMembers().UnsortedList()
-		}
-		for _, ns := range namespaces {
-			pods := c.podsClient.List(ns, klabels.Everything())
-			services := c.services.List(ns, klabels.Everything())
-			c.ambientIndex.HandleSelectedNamespace(ns, pods, services, c)
+		for _, ns := range c.namespaces.List(metav1.NamespaceAll, klabels.Everything()) {
+			pods := c.podsClient.List(ns.Name, klabels.Everything())
+			services := c.services.List(ns.Name, klabels.Everything())
+			c.ambientIndex.HandleSelectedNamespace(ns.Name, pods, services, c)
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:897]: I am not quite sure what you mean by all together?

BTW this sync is for network information changing now only, not namespace (which doesn't need special logic)

**Review by hzxuzhonghu — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

**Pull request merged**